src/audio/smart_amp/smart_amp_maxim_dsm.c

```diff
@@ -0,0 +1,593 @@
+{
+	int *circularbuffersize = hspk->circularbuffersize;
+	int *delayedsamples = hspk->delayedsamples;
+	struct dsm_api_init_ext_t initparam;
+	enum DSM_API_MESSAGE retcode;
+
+	comp_dbg(dev, BUILD_TIME);
+
+	initparam.isamplebitwidth       = 16;
+	initparam.ichannels             = 2;
+	initparam.ipcircbuffersizebytes = circularbuffersize;
+	initparam.ipdelayedsamples      = delayedsamples;
+	initparam.isamplingrate         = 48000;
+
+	retcode = dsm_api_init(hspk->dsmhandle, &initparam,
+			       sizeof(struct dsm_api_init_ext_t));
+	if (retcode != DSM_API_OK) {
+		goto exit;
+	} else	{
+		hspk->ff_fr_sz_samples =
+			initparam.off_framesizesamples;
+		hspk->fb_fr_sz_samples =
+			initparam.ofb_framesizesamples;
+		hspk->channelmask = 0;
+		hspk->nchannels = initparam.ichannels;
+		hspk->ifsamples = hspk->ff_fr_sz_samples
+			* initparam.ichannels;
+		hspk->ibsamples = hspk->fb_fr_sz_samples
+			* initparam.ichannels;
+	}
+
+	comp_dbg(dev, "[DSM] Initialization completed. (module:%p, dsm:%p)",
+		 (uintptr_t)hspk,
+		 (uintptr_t)hspk->dsmhandle);
+
+	return 0;
+exit:
+	comp_err(dev, "[DSM] Initialization failed. ret:%d", retcode);
+	return (int)retcode;
+}
+
+#ifdef PRINT_ADAPTIVE_LOG
+static int maxim_dsm_get_adaptive_apram(struct smart_amp_mod_struct_t *hspk,
+					struct comp_dev *dev)
+{
+	enum DSM_API_MESSAGE retcode;
+	int cmdblock[1 + SMART_AMP_FF_MAX_CH_NUM];
+	int excur[2], temp[2], dc_res[2], fc[2], q[2];
+
+	cmdblock[0] = DSM_SET_CMD_ID(20);
+	retcode = dsm_api_get_params(hspk->dsmhandle, 1, (void *)cmdblock);
+	if (retcode != DSM_API_OK)
+		goto err;
+	excur[0] = cmdblock[1];
+	excur[1] = cmdblock[2];
+
+	cmdblock[0] = DSM_SET_CMD_ID(19);
+	retcode = dsm_api_get_params(hspk->dsmhandle, 1, (void *)cmdblock);
+	if (retcode != DSM_API_OK)
+		goto err;
+	temp[0] = cmdblock[1];
+	temp[1] = cmdblock[2];
+
+	cmdblock[0] = DSM_SET_CMD_ID(18);
+	retcode = dsm_api_get_params(hspk->dsmhandle, 1, (void *)cmdblock);
+	if (retcode != DSM_API_OK)
+		goto err;
+	dc_res[0] = cmdblock[1];
+	dc_res[1] = cmdblock[2];
+
+	cmdblock[0] = DSM_SET_CMD_ID(16);
+	retcode = dsm_api_get_params(hspk->dsmhandle, 1, (void *)cmdblock);
+	if (retcode != DSM_API_OK)
+		goto err;
+	fc[0] = cmdblock[1];
+	fc[1] = cmdblock[2];
+
+	cmdblock[0] = DSM_SET_CMD_ID(17);
+	retcode = dsm_api_get_params(hspk->dsmhandle, 1, (void *)cmdblock);
+	if (retcode != DSM_API_OK)
+		goto err;
+	q[0] = cmdblock[1];
+	q[1] = cmdblock[2];
+
+	comp_dbg(dev, "[DSM] excursion(Q27) :%08x, %08x",
+		 excur[0], excur[1]);
+	comp_dbg(dev,
+		 "[DSM] temp(Q19) :%08x, %08x, dc_res(Q27) :%08x, %08x",
+		 temp[0], temp[1], dc_res[0], dc_res[1]);
+	comp_dbg(dev,
+		 "[DSM] fc(Q9) :%08x, %08x, q(Q29) :%08x, %08x",
+		 fc[0], fc[1], q[0], q[1]);
+
+	return 0;
+err:
+	comp_err(dev, "[DSM] Get adaptive parameters. error:%d", retcode);
+	return retcode;
+}
+#endif
+
+static inline void *_memmove(void *dest, const void *src, size_t count)
```

Comment:
This should be a separate PR for memmove. It can be added to src/lib/lib.c

Reply:
I added memmove function to lib.c on the separate [commit](https://github.com/thesofproject/sof/pull/3248/commits/ade5f9b41144899bb1d72653690d8fe7b7906c23) on the other [PR](https://github.com/thesofproject/sof/pull/3248/) to the main stream.
Let me includes other two [commits](https://github.com/thesofproject/sof/pull/3248/commits) into this PR and create one more commit only for the static DSM library support on the top of two commits.

Reply:
Removed memmove from this PR.